pkg/autodiscovery/common/utils/annotations.go

```diff
@@ -0,0 +1,368 @@
+	if checkNamesJSON, found := annotations[fmt.Sprintf(v1PodAnnotationCheckNamesFormat, adIdentifier)]; found {
+		checkNames, err := ParseCheckNames(checkNamesJSON)
+		if err != nil {
+			return nil, fmt.Errorf("Cannot parse check names: %w", err)
+		}
+
+		return checkNames, nil
+	}
+
+	// AD annotations legacy: "service-discovery.datadoghq.com/redis.check_names"
+	if checkNamesJSON, found := annotations[fmt.Sprintf(legacyPodAnnotationCheckNamesFormat, adIdentifier)]; found {
+		checkNames, err := ParseCheckNames(checkNamesJSON)
+		if err != nil {
+			return nil, fmt.Errorf("Cannot parse check names: %w", err)
+		}
+
+		return checkNames, nil
+	}
+
+	return nil, nil
+}
+
+// ExtractTemplatesFromAnnotations looks for autodiscovery configurations in
+// a map of annotations and returns them if found. In order of priority, it
+// prefers annotations v2, v1, and legacy.
+func ExtractTemplatesFromAnnotations(entityName string, annotations map[string]string, adIdentifier string) ([]integration.Config, []error) {
+	var (
+		configs []integration.Config
+		errors  []error
+		prefix  string
+	)
+
+	if checksJSON, found := annotations[fmt.Sprintf(v2PodAnnotationChecksFormat, adIdentifier)]; found {
+		// AD annotations v2: "ad.datadoghq.com/redis.checks"
+		prefix = fmt.Sprintf(podAnnotationFormat, adIdentifier)
+		c, err := parseChecksJSON(entityName, checksJSON)
+		if err != nil {
+			errors = append(errors, err)
+		} else {
+			configs = append(configs, c...)
+		}
+	} else {
+		// AD annotations v1: "ad.datadoghq.com/redis.check_names"
+		// AD annotations legacy: "service-discovery.datadoghq.com/redis.check_names"
+		prefix = findPrefix(annotations, adPrefixFormats, adIdentifier, checkNamePath)
+		if prefix != "" {
+			c, err := extractCheckTemplatesFromMap(entityName, annotations, prefix)
+			if err != nil {
+				errors = append(errors, fmt.Errorf("could not extract checks config: %v", err))
+			} else {
+				configs = append(configs, c...)
+			}
+		}
+	}
+
+	// prefix might not have been detected if there are no check
+	// annotations, so we try to find a prefix for log configs
+	if prefix == "" {
+		// AD annotations v1: "ad.datadoghq.com/redis.logs"
+		// AD annotations legacy: "service-discovery.datadoghq.com/redis.logs"
+		prefix = findPrefix(annotations, adPrefixFormats, adIdentifier, logsConfigPath)
+	}
+
+	if prefix != "" {
+		c, err := extractLogsTemplatesFromMap(entityName, annotations, prefix)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("could not extract logs config: %v", err))
+		} else {
+			configs = append(configs, c...)
+		}
+	}
+
+	return configs, errors
+}
+
+// parseChecksJSON parses an AD annotation v2
+// (ad.datadoghq.com/redis.check_names) JSON string into []integration.Config.
+func parseChecksJSON(adIdentifier string, checksJSON string) ([]integration.Config, error) {
+	var namedChecks map[string]struct {
+		Name       string           `json:"name"`
+		InitConfig integration.Data `json:"init_config"`
```

Comment:
shouldn't this be of type `*integration.Data` to mark it as optional field? https://go.dev/play/p/1oxDEprdUG0